### ISSUE TITLE
Entire CSS output in entry in when splitting

### ISSUE THREAD

**Issue by maisano** [evanw/esbuild#1590]
First, thanks for your work on `esbuild`–the experience of using it is quite refreshing.

In endeavoring to port the build system of a very large app over to esbuild I had noticed that the entry point CSS output contains the CSS of every module in the graph, even when the `splitting` option is set to `true`. The application in question has a very slim entry point, with a small JS runtime and a few styled elements–the bulk of the code lies in its split chunks. As you would likely guess, the resultant CSS for the entry point itself ends up being enormous.

Is this intentional, or perhaps a known issue or limitation? I poked around a bit and didn't notice any obvious references to this.

**Comment by evanw** [comment]
This is by design. I believe that trying to split CSS code while still generating output files that are CSS (instead of JS) could result in visual differences when that code loads because CSS is order-dependent, which would cause bugs. Having each entry point have its own CSS output file containing all CSS reachable from that entry point avoids such bugs.

Specifically, there are two things that code splitting does: share code between entry points and enable lazy-loading of code due to dynamic entry points from `import()` expressions. It's technically possible to generate a CSS output file with the correct order and with code splitting for the first thing (in the limit nothing would be bundled and the output directory would just recreate the input directory with individual CSS files), but I'm not sure it's possible to have dynamic `import()` entry points use CSS output files and still get the CSS order correct. The problem is that the CSS used by the dynamic entry point may need to be inserted in between sections of already-loaded CSS, but adding a new CSS style element just appends it to the end.

If I recall correctly, I believe Webpack solves this by outputting CSS as strings in JS instead and then dynamically editing a CSS style element at run-time as new chunks are loaded. But I didn't want to do that approach with esbuild because there are good loading-time benefits to having CSS code be in a separate CSS file instead of embedded in the JS file.

**Comment by maisano** [comment]
Makes sense–thanks for the clarification here. Some additional context is that we're using CSS modules, which certainly helps us be less encumbered by ordering issues. Very broadly, we avoid leaky styles (anything non-scoped) and isolate them to stylesheets external to split chunks. We also avoid `@import` rules in any of the CSS modules to the same effect.

While I'm sure there are usages that could go sideways here, I'm curious if adding this functionality behind a flag is tenable. The chunked CSS is still output correctly–in my quick testing, manually trimming the entry output CSS from the first chunk reference hasn't resulted in any style issues.

**Comment by maisano** [comment]
I had some time to dig into other approaches yesterday and I landed on creating a plugin that more or less emulates the behavior of webpack's `MiniCssExtractPlugin`. This works by doing an initial build of the entry points with splitting enabled and all CSS imports marked as external. Something like:

```js
build({
  entryPoints: ['main', 'etc'],
  external: ['*.css'],
  bundle: true,
  format: 'esm',
  splitting: true,
})
```

There's then a second pass on the output JS files that marks every non-CSS dependency as external, which ultimately processes and emits a single sibling CSS file for any JS asset that contains a CSS import. Since all CSS in the graph uses CSS modules, there aren't any collisions or errant styles due to load / insert order. As I mentioned previously, there are _some_ global rules, but those live external to any chunk code. So far this works like charm.

One downside to this approach is that there's no obvious place to do the style injection–in my prototyping, the plugin inlines a runtime to relevant JS modules in order to do the injection, however, it would be more ideal to trigger this upstream, nearer the position of the dynamic import expression (to avoid a waterfall of JS -> CSS loading). Curious if you have thoughts on this, or the model I've outlined above.

Otherwise, are you still planning on having esbuild natively support CSS modules?

**Comment by threepointone** [comment]
### Making the case for implementing code splitting without worrying much about insertion order: 


Consider the different scenarios where this would apply - 

- For folks who are using dynamic `import()` statements, they are _already aware_ that they might not get the right insertion order, and will lean on solutions like css modules (which hashes classnames and doesn't rely on the cascade).
- And if they _do_ want proper insertion order, they will implement a webpack style plugin that inlines the css file as text into their js bundle.
- And for the folks who would rather have a single css file, they will not enable the `--splitting` flag. 

Could I suggest then, that instead of `--splitting` doing both js and css code splitting, would you have separate flags for js and css? (Perhaps call the additional one `--css-splitting`). So folks could opt into one of the above 3 behaviours. Calling `import('./x.js')` which possibly has an `import 'x.css';` statement, would then load the the two of them in parallel (instead of the waterfall that @maisano describes).

Is this reasonable?

**Comment by mhsdesign** [comment]
I was interested in prepacking/building a library and the copy loader for css works wonders ;) (it even works for css modules if you use a plugin that supports it like https://github.com/mhsdesign/esbuild-composes-from-css-modules)
